File: camerax_android/lib/src/core/camera_selector_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class CameraSelector$BuilderImpl implements CameraSelector$Builder {
  final CameraSelectorBuilderProxyApi api;

  CameraSelector$BuilderImpl.internal(this.api);

  @override
  Future<CameraSelector$Builder> addCameraFilter(CameraFilter cameraFilter) =>
      api.addCameraFilter(cameraFilter.api).then((e) => e.impl);

  @override
  Future<CameraSelector> build() => api.build().then((e) => e.impl);

  @override
  Future<CameraSelector$Builder> requireLensFacing(
    CameraSelector$LensFacing lensFacing,
  ) => api.requireLensFacing(lensFacing.api).then((e) => e.impl);

  @override
  Future<CameraSelector$Builder> setPhysicalCameraId(String physicalCameraId) =>
      api.setPhysicalCameraId(physicalCameraId).then((e) => e.impl);
}

final class CameraSelectorImpl implements CameraSelector {
  final CameraSelectorProxyApi api;

  CameraSelectorImpl.internal(this.api);

  @override
  Future<List<CameraInfo>> filter(List<CameraInfo> cameraInfos) => api
      .filter(cameraInfos.map((e) => e.api).toList())
      .then((e) => e.map((e1) => e1.impl).toList());

  @override
  Future<String?> getPhysicalCameraId() => api.getPhysicalCameraId();
}

final class CameraSelectorChannelImpl extends CameraSelectorChannel {
  @override
  CameraSelector get back =>
      CameraSelectorImpl.internal(CameraSelectorProxyApi.back);
  @override
  CameraSelector get front =>
      CameraSelectorImpl.internal(CameraSelectorProxyApi.front);
  @override
  CameraSelector get external =>
      CameraSelectorImpl.internal(CameraSelectorProxyApi.external);

  @override
  CameraSelector$Builder createBuilder() {
    final api = CameraSelectorBuilderProxyApi();
    return CameraSelector$BuilderImpl.internal(api);
  }
}

extension CameraSelector$LensFacingX on CameraSelector$LensFacing {
  CameraSelectorLensFacingApi get api =>
      CameraSelectorLensFacingApi.values[index];
}

extension CameraSelectorLensFacingApiX on CameraSelectorLensFacingApi {
  CameraSelector$LensFacing get impl => CameraSelector$LensFacing.values[index];
}

extension CameraSelectorBuilderProxyApiX on CameraSelectorBuilderProxyApi {
  CameraSelector$Builder get impl => CameraSelector$BuilderImpl.internal(this);
}

extension CameraSelectorX on CameraSelector {
  CameraSelectorProxyApi get api {
    final impl = this;
    if (impl is! CameraSelectorImpl) throw TypeError();
    return impl.api;
  }
}

extension CameraSelectorProxyApiX on CameraSelectorProxyApi {
  CameraSelector get impl => CameraSelectorImpl.internal(this);
}

File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/core/CameraSelectorImpl.kt
package dev.zeekr.camerax_android.core

import androidx.annotation.OptIn
import androidx.camera.core.CameraFilter
import androidx.camera.core.CameraInfo
import androidx.camera.core.CameraSelector
import androidx.camera.core.ExperimentalLensFacing
import dev.zeekr.camerax_android.CameraSelectorLensFacingApi
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.PigeonApiCameraSelectorBuilderProxyApi
import dev.zeekr.camerax_android.PigeonApiCameraSelectorProxyApi

class CameraSelectorImpl(registrar: CameraXApiPigeonProxyApiRegistrar) : PigeonApiCameraSelectorProxyApi(registrar) {
    class BuilderImpl(registrar: CameraXApiPigeonProxyApiRegistrar) :
        PigeonApiCameraSelectorBuilderProxyApi(registrar) {
        override fun pigeon_defaultConstructor(): CameraSelector.Builder {
            return CameraSelector.Builder()
        }

        override fun addCameraFilter(
            pigeon_instance: CameraSelector.Builder, cameraFilter: CameraFilter
        ): CameraSelector.Builder {
            return pigeon_instance.addCameraFilter(cameraFilter)
        }

        override fun requireLensFacing(
            pigeon_instance: CameraSelector.Builder, lensFacing: CameraSelectorLensFacingApi
        ): CameraSelector.Builder {
            return pigeon_instance.requireLensFacing(lensFacing.impl)
        }

        override fun setPhysicalCameraId(
            pigeon_instance: CameraSelector.Builder, physicalCameraId: String
        ): CameraSelector.Builder {
            return pigeon_instance.setPhysicalCameraId(physicalCameraId)
        }

        override fun build(pigeon_instance: CameraSelector.Builder): CameraSelector {
            return pigeon_instance.build()
        }
    }

    override fun back(): CameraSelector {
        return CameraSelector.DEFAULT_BACK_CAMERA
    }

    override fun front(): CameraSelector {
        return CameraSelector.DEFAULT_FRONT_CAMERA
    }

    @OptIn(ExperimentalLensFacing::class)
    override fun external(): CameraSelector {
        return CameraSelector.Builder().requireLensFacing(CameraSelector.LENS_FACING_EXTERNAL).build()
    }

    override fun filter(pigeon_instance: CameraSelector, cameraInfos: List<CameraInfo>): List<CameraInfo> {
        return pigeon_instance.filter(cameraInfos)
    }

    override fun getPhysicalCameraId(pigeon_instance: CameraSelector): String? {
        return pigeon_instance.physicalCameraId
    }
}

val Int.cameraSelectorLensFacingApi: CameraSelectorLensFacingApi
    @OptIn(ExperimentalLensFacing::class) get() = when (this) {
        CameraSelector.LENS_FACING_UNKNOWN -> CameraSelectorLensFacingApi.UNKNOWN
        CameraSelector.LENS_FACING_FRONT -> CameraSelectorLensFacingApi.FRONT
        CameraSelector.LENS_FACING_BACK -> CameraSelectorLensFacingApi.BACK
        CameraSelector.LENS_FACING_EXTERNAL -> CameraSelectorLensFacingApi.EXTERNAL
        else -> throw NotImplementedError("Not implemented value: $this")
    }

val CameraSelectorLensFacingApi.impl: Int
    @OptIn(ExperimentalLensFacing::class) get() = when (this) {
        CameraSelectorLensFacingApi.UNKNOWN -> CameraSelector.LENS_FACING_UNKNOWN
        CameraSelectorLensFacingApi.FRONT -> CameraSelector.LENS_FACING_FRONT
        CameraSelectorLensFacingApi.BACK -> CameraSelector.LENS_FACING_BACK
        CameraSelectorLensFacingApi.EXTERNAL -> CameraSelector.LENS_FACING_EXTERNAL
    }
